driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -86,38 +92,59 @@ public void run() {
             }
         };
 
+        this.connections = new CopyOnWriteArrayList<PooledConnection>();
+        this.open = new AtomicInteger();
+
         // Create initial core connections
-        List<PooledConnection> l = new ArrayList<PooledConnection>(options().getCoreConnectionsPerHost(hostDistance));
-        try {
-            for (int i = 0; i < options().getCoreConnectionsPerHost(hostDistance); i++)
-                l.add(manager.connectionFactory().open(this));
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            // If asked to interrupt, we can skip opening core connections, the pool will still work.
-            // But we ignore otherwise cause I'm not sure we can do much better currently.
-        } catch (ConnectionException e) {
-            forceClose(l);
-            throw e;
-        } catch (UnsupportedProtocolVersionException e) {
-            forceClose(l);
-            throw e;
-        } catch (ClusterNameMismatchException e) {
-            forceClose(l);
-            throw e;
-        } catch (RuntimeException e) {
-            forceClose(l);
-            throw e;
-        }
-        this.connections = new CopyOnWriteArrayList<PooledConnection>(l);
-        this.open = new AtomicInteger(connections.size());
+        final List<ListenableFuture<PooledConnection>> connectionFutures =
+            Lists.newArrayListWithCapacity(options().getCoreConnectionsPerHost(hostDistance));
+        for (int i = 0; i < options().getCoreConnectionsPerHost(hostDistance); i++)
+            connectionFutures.add(manager.connectionFactory().openAsync(this));
 
-        logger.trace("Created connection pool to host {}", host);
+        ListenableFuture<List<PooledConnection>> allConnectionsFuture = Futures.allAsList(connectionFutures);
+
+        Futures.addCallback(allConnectionsFuture,
+            new FutureCallback<List<PooledConnection>>() {
+                @Override
+                public void onSuccess(List<PooledConnection> l) {
+                    connections.addAll(l);
+                    open.set(l.size());
```

Comment:
One small issue here is that there is a small window where the Pool future is marked as complete, but open is not set yet.   This could cause issues in the various places we check open when spawning and trashing connections.  I've seen this a few times in SessionTest#sessionMemoryLeak test, but it's very hard to reproduce.  I think we need another intermediate future that gets set after open.set completes in onSuccess that initFuture transforms off of.